src/main/java/uk/nhs/hee/trainee/details/api/DspCredentialResource.java

```diff
@@ -0,0 +1,267 @@
+   * @param parEndpoint                the DSP Pushed Authorization Request (PAR) endpoint
+   * @param authorizeEndpoint          the DSP credential Authorize Request endpoint
+   * @param tokenEndpoint              the DSP Token Request endpoint
+   * @param restTemplateBuilder        the REST template builder
+   * @param jwtService                 the JWT service
+   */
+  public DspCredentialResource(PlacementService placementService,
+                               ProgrammeMembershipService programmeMembershipService,
+                               ObjectMapper objectMapper,
+                               @Value("${dsp.client-id}") String clientId,
+                               @Value("${dsp.client-secret}") String clientSecret,
+                               @Value("${dsp.redirect-uri}") String redirectUri,
+                               @Value("${dsp.par-endpoint}") String parEndpoint,
+                               @Value("${dsp.authorize-endpoint}") String authorizeEndpoint,
+                               @Value("${dsp.token.issue-endpoint}") String tokenEndpoint,
+                               RestTemplateBuilder restTemplateBuilder,
+                               JwtService jwtService) {
+    this.placementService = placementService;
+    this.programmeMembershipService = programmeMembershipService;
+    this.objectMapper = objectMapper;
+    this.clientId = clientId;
+    this.clientSecret = clientSecret;
+    this.redirectUri = redirectUri;
+    this.parEndpoint = parEndpoint;
+    this.issueEndpoint = authorizeEndpoint;
+    this.tokenEndpoint = tokenEndpoint;
+    this.restTemplateBuilder = restTemplateBuilder;
+    this.jwtService = jwtService;
+  }
+
+  /**
+   * Get the PAR response, including the request URI. We assume the trainee has been verified. This
+   * would be called by the front-end when the user clicks on the 'Add to wallet' button for a
+   * placement or programme membership.
+   *
+   * <p>NOTE: in future we will pass the content from the front-end as a JWT so that we can verify
+   * that it has not been tampered with.
+   *
+   * @param credentialType 'placement' or 'programmemembership'
+   * @param tisId The ID of the placement / programme membership.
+   * @return The PAR response, or Internal Server Error if the request times out
+   */
+  @GetMapping(value = "/par/{credentialType:placement|programmemembership}/{tisId}")
+  public ResponseEntity<ParResponse> getCredentialParUri(
+      @PathVariable(name = "credentialType") String credentialType,
+      @PathVariable(name = "tisId") String tisId,
+      @RequestHeader(HttpHeaders.AUTHORIZATION) String token) {
+    String traineeTisId;
+    log.info("Get credential request URI for {} with TIS ID {}", credentialType, tisId);
+
+    String[] tokenSections = token.split("\\.");
+    byte[] payloadBytes = Base64.getUrlDecoder()
+        .decode(tokenSections[1].getBytes(StandardCharsets.UTF_8));
```

Comment:
We don't verify this token (to do this it looks like we'd need the public JSON Web Key (JWK) for the cognito user pool). This might be ok since the endpoint is not publicly visible? I was worried a logged-in user could retrieve any other user's profile by manipulating the auth token)

Reply:
The API Gateway verifies the token is valid and unmodified, so we can assume this is a safe approach.